### PR TITLE
Fixed RA2 lobrdb sprites not getting loaded

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/12115.

### CODE REVIEW

File: OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs [hunk 1]
```diff
@@ -118,7 +118,9 @@ bool IsShpTS(Stream s)
 			while (w == 0 && h == 0 && f++ < imageCount);
 
 			s.Position = start;
-			return type < 4;
+
+			// Type 18 is a special case for RA2 low bridges.
```
Comment (pchote) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: I'm sorry to say that i'm pretty sure that this is not a special case, but the result of some other bug that still hasn't been found and needs to be fixed.

https://github.com/zzattack/ccmaps-net/blob/master/CNCMaps.FileFormats/ShpFile.cs#L58 does not have any special handing for a type 18, and that project has no issues with the RA2 lobrdb sprites.

Comment (Mailaender) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: As far as I see it that parser doesn't have the "Check the size and format flag" our loader has. I didn't know that the type means compression type. For the first 3 frames of a RA2 low bridge tile I debugged the following

```
lobrdb01.sno
w 0 h 0 type 4
w 0 h 0 type 0
w 0 h 64256 type 18
```

with actual data being in the 2nd frame so it seems to be type 0 (uncompressed data) with no width/height which we discard as invalid. The 3rd frame has a gigantic height and my so called type 18 so I guess other data is stored in there.

Can I just remove the `w == 0 && h == 0` check or can you maybe explain for which files you added it so we can change this exceptional case?

Comment (pchote) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: The idea is to skip bogus frames, which until now have had *both* width and height equal to zero.  I expect that changing that `&&` to an `||` should fix the issue here.

Comment (pchote) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: The downside of course is that it raises the chances of falsely identifying non-shp images with unfortunate coincidences in byte patterns as shp.

Comment (Mailaender) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: Sadly the same case here:
```
w 0 h 0 type 0
```

is the actual frame we want and it has _both_ width and height equal to zero.

Looping through all frames:
```
lobrdb01.sno
w 0 h 0 type 4
w 0 h 0 type 0
w 0 h 64256 type 18
w 0 h 0 type 0
w 30 h 19 type 120
```
none will pass the test even if we allow `w == 0 || h == 0` but don't accept `type == 18`.

Comment (pchote) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: Are you sure that the test method is being called with the correct data (and the correct stream position)?  30,19 doesn't look like a reasonable frame size for a 3x1 diagonal bridge tile, so I suspect you may be parsing random bytes.

Comment (pchote) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: https://github.com/OpenRA/OpenRA/blob/287290bbdf08ac1ec884213ef1612a185242c2a1/OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs#L108-L118

A quick check against ra2ff.txt confirms that this is completely bogus, and tries to parse unrelated data as frame width/height/compression types.  This loop needs to be rewritten to account for the actual layout of the image header.

Comment (Mailaender) [inline OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs]: Updated as discussed in IRC.

**Comment by GraionDilach** [comment]
Is this intends to fix the "OpenRA-calls-SHP(TS)-files-invalid-if-they-start-with-an-empty-frame" bug?

**Comment by Mailaender** [comment]
Maybe. The RA2 low bridge tiles come with an empty frame and will now be detected as valid ShpTS:

```
lobrdb01.sno
w 0 h 0 type 4
w 120 h 70 type 3
```

while it was unknown bits read before, as we didn't actually scan the image headers.

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
> Is this intends to fix the "OpenRA-calls-SHP(TS)-files-invalid-if-they-start-with-an-empty-frame" bug?

The current version of the PR should fix that, yes.

**Comment by reaperrr** [comment]
Can someone provide a test-case shp? I don't feel like downloading RA2 just for this (I have it on Origin, but not currently installed).

**Comment by GraionDilach** [comment]
infshdw as the RA2 Rocketeer shadow for first-frame-empty bug, ngmisl is for the all-frames-empty extreme case.

[shptstest.zip](https://github.com/OpenRA/OpenRA/files/1290580/shptstest.zip)

**Comment by reaperrr** [comment]
Thanks.

`infshdw` works fine with this, `ngmisl` still triggers a "tried to read beyond stream" error:
https://gist.github.com/reaperrr/4ca14b9dfd8303659fa9e913f5a10965

Since it's an entirely empty sprite, I'd say as long as it doesn't crash the game and just silently logs an error, it's probably fine.

I'll wait for @pchote's opinion before merging this.

**Comment by Mailaender** [comment]
You can also test via https://github.com/OpenRA/ra2/pull/393 which references this pull request. If you translate https://github.com/OpenRA/ra2/blob/master/fetch-content.sh into PowerShell, it should be convenient to test as well without Origin or CD drives.

**Comment by pchote** [comment]
> Since it's an entirely empty sprite, I'd say as long as it doesn't crash the game and just silently logs an error, it's probably fine.

The code is actually still wrong, the `f++` should be `++f` for the loop to terminate at the right time.
This then leads to a second set of crashes from frames that are completely empty but which we still use ingame for some reason (null.shp, lobrdg28.tem).

**Comment by pchote** [comment]
Following on with this... We can avoid the crash I refer to above by being a bit smarter about the checking.  If we are able to identify the frame types that are used with zero-sized sprites then we can have an explicit check to allow these all-empty sprites to be accepted without false-matching against all-zero data.

**Comment by Mailaender** [comment]
Updated and confirmed that #393 still works as well as the RA2 `infshdw` and `ngmisl` as well as TS `null.shp` and `lobrdg28`.

**Review by reaperrr — APPROVED** [review]
Works with all test files now 👍

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3bf776b80b1f16cd4d83f9c1bb468ae71e7af1ef...1b25d6b7b83d551e518f74b36de1543fc396ee3e)

**Pull request merged**